### ISSUE TITLE
Undefined plan with nthreads. This is a bug

### ISSUE THREAD

**Issue by masip85** [pyFFTW/pyFFTW#294]
I am obtaining this error:
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
RuntimeError: Undefined plan with nthreads. This is a bug

Exception ignored in: 'pyfftw.pyfftw._fftw_plan_with_nthreads_null'
RuntimeError: Undefined plan with nthreads. This is a bug

BUT it seems process ends, but much much more slow than scipy fftpocket.

I have compiled fftw with:
    ./configure --enable-single --enable-shared --enable-neon ARM_CPU_TYPE=cortex-a8 ; \

Is there anyway to check if it is using SMID?

**Comment by hgomersall** [comment]
Oh, this is actually a bug! The point is the error should have been caught before it got to that line. Excitingly, it's the first time that line has ever been triggered.

There are compile flags `HAVE_SINGLE`, `HAVE_DOUBLE` and `HAVE_LONG`, as well as `HAVE_SINGLE_MULTITHREADING`, `HAVE_DOUBLE_MULTITHREADING` and `HAVE_LONG_MULTITHREADING`.

These flags are set at compile time in `setup.py`. The problem is that they need to agree, so `HAVE_SINGLE_MULTITHREADING` must be true if `HAVE_SINGLE` is true and equivalently for the other two. Arguably, this is not the right policy, but it does allow the clear bug pointer you've highlighted.

The line it is set is [here](https://github.com/pyFFTW/pyFFTW/blob/master/setup.py#L292). It would be good to know why it's not populating `compile_time_env` properly. Are you able to poke around a bit to get some more info?

**Comment by masip85** [comment]
I would be able with a little bit of help or introduction of what you would like me to do. Because I get that error,and that's all. :S

**Comment by hgomersall** [comment]
`setup.py` is run on build. A starting point would be to know what the compile time options are used, which should be populated in [this loop](https://github.com/pyFFTW/pyFFTW/blob/ff5e19f5aedeb52658f2ea6b3fa5541849d77e1e/setup.py#L246). You can stick some prints in to see how `self.compile_time_env` gets updated as the loop iterates. What should happen is those flags described above should get populated properly. It seems like the build system can't work out what threading library it can use.

To confirm, does it work if you don't try to set threads > 1?

**Comment by masip85** [comment]
Ok,I can give it a try, until now,I used pip to install. To add that print I should modify this code, and build manually this repo like this,isn't it?:
`python setup.py build_ext --inplace`

It shouldn't be able to enable threads,because I didn't compile with that option. In fact, 

_ValueError: threads > 1 requested, but pyFFTW was not built with multithreaded FFTW._

But look,if interface is cached error dissapears:
![imagen](https://user-images.githubusercontent.com/12863854/87028588-1e649f00-c1df-11ea-8611-fd123a5f9cd0.png)

But rarely still slower than fftpocket, how to check sdim is working? Maybe I should try also do this directly in C.

**Comment by masip85** [comment]
By the way:
```
import pyfftw
a = pyfftw.empty_aligned(4, dtype='complex128')
fft = pyfftw.builders.fft(a)
a[:] = [1, 2, 3, 4]
fft() #
array([0.+0.j, 0.+0.j, 0.+0.j, 0.+0.j], dtype=complex64)
```
not 
_array([ 10.+0.j,  -2.+2.j,  -2.+0.j,  -2.-2.j])_

**Comment by hgomersall** [comment]
That seems wrong to me.

Your assessment of how to build manually is correct.

**Comment by masip85** [comment]
I added:
```
for d, s in zip(data_types, data_types_short):
            print('DEBUG - - - - compile_time_env:')
            print(self.compile_time_env)
```
It looks like doesn't get into the loop.

```
Step 17/19 : RUN /bin/bash -c '     cd /etc/pyFFTW && python setup.py build_ext --inplace;     '
 ---> Running in 16e467eb2907
running build_ext
Build pyFFTW with support for FFTW with
single precision
cythoning pyfftw/pyfftw.pyx to pyfftw/pyfftw.c
building 'pyfftw.pyfftw' extension
creating build
creating build/temp.linux-aarch64-3.7
creating build/temp.linux-aarch64-3.7/pyfftw
gcc -pthread -B /opt/conda/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/conda/include/python3.7m -I/etc/pyFFTW/include -I/etc/pyFFTW/pyfftw -I/opt/conda/lib/python3.7/site-packages/numpy/core/include -I/opt/conda/include -I/opt/conda/include/python3.7m -c pyfftw/pyfftw.c -o build/temp.linux-aarch64-3.7/pyfftw/pyfftw.o
In file included from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1832:0,
                 from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from pyfftw/pyfftw.c:606:
/opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it with " \
  ^~~~~~~
pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw_byte_align.isra.64.constprop’:
pyfftw/pyfftw.c:3969:6: warning: ‘__pyx_v_update_dtype’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   if (__pyx_t_1) {
      ^
creating build/lib.linux-aarch64-3.7
creating build/lib.linux-aarch64-3.7/pyfftw
gcc -pthread -shared -B /opt/conda/compiler_compat -L/opt/conda/lib -Wl,-rpath=/opt/conda/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-aarch64-3.7/pyfftw/pyfftw.o -L/opt/conda/lib -lfftw3f -lm -o build/lib.linux-aarch64-3.7/pyfftw/pyfftw.cpython-37m-aarch64-linux-gnu.so
copying build/lib.linux-aarch64-3.7/pyfftw/pyfftw.cpython-37m-aarch64-linux-gnu.so -> pyfftw
/opt/conda/lib/python3.7/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /etc/pyFFTW/pyfftw/pyfftw.pxd
  tree = Parsing.p_module(s, pxd, full_module_name)
warning: pyfftw/pyfftw.pyx:344:22: Casting a GIL-requiring function into a nogil function circumvents GIL validation
warning: pyfftw/pyfftw.pyx:364:23: Casting a GIL-requiring function into a nogil function circumvents GIL validation
Removing intermediate container 16e467eb2907
 ---> 99bf76624270
```

**Comment by hgomersall** [comment]
It must be - possible the print isn't actually outputting to stdout. Not sure how to fix that...

**Comment by masip85** [comment]
log.info was the answer:
```
RUN cd /etc/pyFFTW && python setup.py build_ext --inplace;
 ---> Running in 88881df96417
running build_ext
DEBUG - - - - Data types:
['DOUBLE', 'SINGLE', 'LONG']
DEBUG - - - - Data types short:
['', 'f', 'l']
DEBUG - - - - compile_time_env:
{}
DEBUG - - - - compile_time_env:
{'HAVE_DOUBLE': False, 'HAVE_DOUBLE_OMP': False, 'HAVE_DOUBLE_THREADS': False, 'HAVE_DOUBLE_MULTITHREADING': False, 'HAVE_DOUBLE_MPI': False}
DEBUG - - - - compile_time_env:
{'HAVE_DOUBLE': False, 'HAVE_DOUBLE_OMP': False, 'HAVE_DOUBLE_THREADS': False, 'HAVE_DOUBLE_MULTITHREADING': False, 'HAVE_DOUBLE_MPI': False, 'HAVE_SINGLE': True, 'HAVE_SINGLE_OMP': False, 'HAVE_SINGLE_THREADS': False, 'HAVE_SINGLE_MULTITHREADING': False, 'HAVE_SINGLE_MPI': False}
Build pyFFTW with support for FFTW with
single precision
cythoning pyfftw/pyfftw.pyx to pyfftw/pyfftw.c
building 'pyfftw.pyfftw' extension
creating build
creating build/temp.linux-aarch64-3.7
creating build/temp.linux-aarch64-3.7/pyfftw
gcc -pthread -B /opt/conda/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/conda/include/python3.7m -I/etc/pyFFTW/include -I/etc/pyFFTW/pyfftw -I/opt/conda/lib/python3.7/site-packages/numpy/core/include -I/opt/conda/include -I/opt/conda/include/python3.7m -c pyfftw/pyfftw.c -o build/temp.linux-aarch64-3.7/pyfftw/pyfftw.o
In file included from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1832:0,
                 from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from pyfftw/pyfftw.c:606:
/opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it with " \
  ^~~~~~~
pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw_byte_align.isra.64.constprop’:
pyfftw/pyfftw.c:3969:6: warning: ‘__pyx_v_update_dtype’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   if (__pyx_t_1) {
      ^
creating build/lib.linux-aarch64-3.7
creating build/lib.linux-aarch64-3.7/pyfftw
gcc -pthread -shared -B /opt/conda/compiler_compat -L/opt/conda/lib -Wl,-rpath=/opt/conda/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-aarch64-3.7/pyfftw/pyfftw.o -L/opt/conda/lib -lfftw3f -lm -o build/lib.linux-aarch64-3.7/pyfftw/pyfftw.cpython-37m-aarch64-linux-gnu.so
copying build/lib.linux-aarch64-3.7/pyfftw/pyfftw.cpython-37m-aarch64-linux-gnu.so -> pyfftw
/opt/conda/lib/python3.7/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /etc/pyFFTW/pyfftw/pyfftw.pxd
  tree = Parsing.p_module(s, pxd, full_module_name)
warning: pyfftw/pyfftw.pyx:344:22: Casting a GIL-requiring function into a nogil function circumvents GIL validation
warning: pyfftw/pyfftw.pyx:364:23: Casting a GIL-requiring function into a nogil function circumvents GIL validation
Removing intermediate container 88881df96417
 ---> 64e9b30b0550
```

**Comment by hgomersall** [comment]
FYI it would be best to see `compile_time_env` at the conclusion of the loop, but I think we can see that it's not finding any thread support (all the fields are `False`). Can you try `PYFFTW_USE_PTHREADS=1 python setup.py build_ext --inplace` and report back on the logs. I _think_ that might fix it.

**Comment by masip85** [comment]
Now at the end:
```
Step 17/18 : RUN cd /etc/pyFFTW && python setup.py build_ext --inplace;
 ---> Running in b035880cf102
running build_ext
DEBUG - - - - Data types:
['DOUBLE', 'SINGLE', 'LONG']
DEBUG - - - - Data types short:
['', 'f', 'l']
DEBUG - - - - compile_time_env:
{'HAVE_DOUBLE': False, 'HAVE_DOUBLE_OMP': False, 'HAVE_DOUBLE_THREADS': False, 'HAVE_DOUBLE_MULTITHREADING': False, 'HAVE_DOUBLE_MPI': False}
DEBUG - - - - compile_time_env:
{'HAVE_DOUBLE': False, 'HAVE_DOUBLE_OMP': False, 'HAVE_DOUBLE_THREADS': False, 'HAVE_DOUBLE_MULTITHREADING': False, 'HAVE_DOUBLE_MPI': False, 'HAVE_SINGLE': True, 'HAVE_SINGLE_OMP': False, 'HAVE_SINGLE_THREADS': False, 'HAVE_SINGLE_MULTITHREADING': False, 'HAVE_SINGLE_MPI': False}
DEBUG - - - - compile_time_env:
{'HAVE_DOUBLE': False, 'HAVE_DOUBLE_OMP': False, 'HAVE_DOUBLE_THREADS': False, 'HAVE_DOUBLE_MULTITHREADING': False, 'HAVE_DOUBLE_MPI': False, 'HAVE_SINGLE': True, 'HAVE_SINGLE_OMP': False, 'HAVE_SINGLE_THREADS': False, 'HAVE_SINGLE_MULTITHREADING': False, 'HAVE_SINGLE_MPI': False, 'HAVE_LONG': False, 'HAVE_LONG_OMP': False, 'HAVE_LONG_THREADS': False, 'HAVE_LONG_MULTITHREADING': False, 'HAVE_LONG_MPI': False}
Build pyFFTW with support for FFTW with
single precision
cythoning pyfftw/pyfftw.pyx to pyfftw/pyfftw.c
building 'pyfftw.pyfftw' extension
creating build
creating build/temp.linux-aarch64-3.7
creating build/temp.linux-aarch64-3.7/pyfftw
gcc -pthread -B /opt/conda/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/conda/include/python3.7m -I/etc/pyFFTW/include -I/etc/pyFFTW/pyfftw -I/opt/conda/lib/python3.7/site-packages/numpy/core/include -I/opt/conda/include -I/opt/conda/include/python3.7m -c pyfftw/pyfftw.c -o build/temp.linux-aarch64-3.7/pyfftw/pyfftw.o
In file included from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1832:0,
                 from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from pyfftw/pyfftw.c:606:
/opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it with " \
  ^~~~~~~
pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw_byte_align.isra.64.constprop’:
pyfftw/pyfftw.c:3969:6: warning: ‘__pyx_v_update_dtype’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   if (__pyx_t_1) {
      ^
creating build/lib.linux-aarch64-3.7
creating build/lib.linux-aarch64-3.7/pyfftw
gcc -pthread -shared -B /opt/conda/compiler_compat -L/opt/conda/lib -Wl,-rpath=/opt/conda/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-aarch64-3.7/pyfftw/pyfftw.o -L/opt/conda/lib -lfftw3f -lm -o build/lib.linux-aarch64-3.7/pyfftw/pyfftw.cpython-37m-aarch64-linux-gnu.so
copying build/lib.linux-aarch64-3.7/pyfftw/pyfftw.cpython-37m-aarch64-linux-gnu.so -> pyfftw
/opt/conda/lib/python3.7/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /etc/pyFFTW/pyfftw/pyfftw.pxd
  tree = Parsing.p_module(s, pxd, full_module_name)
warning: pyfftw/pyfftw.pyx:344:22: Casting a GIL-requiring function into a nogil function circumvents GIL validation
warning: pyfftw/pyfftw.pyx:364:23: Casting a GIL-requiring function into a nogil function circumvents GIL validation
Removing intermediate container b035880cf102
 ---> 7bfb29f34cc4
```


with use PYFFTW_USE_PTHREADS=1  

```
---> Running in 7565f680b6fd
running build_ext
DEBUG - - - - Data types:
['DOUBLE', 'SINGLE', 'LONG']
DEBUG - - - - Data types short:
['', 'f', 'l']
DEBUG - - - - compile_time_env:
{'HAVE_DOUBLE': False, 'HAVE_DOUBLE_OMP': False, 'HAVE_DOUBLE_THREADS': False, 'HAVE_DOUBLE_MULTITHREADING': False, 'HAVE_DOUBLE_MPI': False}
DEBUG - - - - compile_time_env:
{'HAVE_DOUBLE': False, 'HAVE_DOUBLE_OMP': False, 'HAVE_DOUBLE_THREADS': False, 'HAVE_DOUBLE_MULTITHREADING': False, 'HAVE_DOUBLE_MPI': False, 'HAVE_SINGLE': True, 'HAVE_SINGLE_OMP': False, 'HAVE_SINGLE_THREADS': False, 'HAVE_SINGLE_MULTITHREADING': False, 'HAVE_SINGLE_MPI': False}
DEBUG - - - - compile_time_env:
{'HAVE_DOUBLE': False, 'HAVE_DOUBLE_OMP': False, 'HAVE_DOUBLE_THREADS': False, 'HAVE_DOUBLE_MULTITHREADING': False, 'HAVE_DOUBLE_MPI': False, 'HAVE_SINGLE': True, 'HAVE_SINGLE_OMP': False, 'HAVE_SINGLE_THREADS': False, 'HAVE_SINGLE_MULTITHREADING': False, 'HAVE_SINGLE_MPI': False, 'HAVE_LONG': False, 'HAVE_LONG_OMP': False, 'HAVE_LONG_THREADS': False, 'HAVE_LONG_MULTITHREADING': False, 'HAVE_LONG_MPI': False}
Build pyFFTW with support for FFTW with
single precision
cythoning pyfftw/pyfftw.pyx to pyfftw/pyfftw.c
building 'pyfftw.pyfftw' extension
creating build
creating build/temp.linux-aarch64-3.7
creating build/temp.linux-aarch64-3.7/pyfftw
gcc -pthread -B /opt/conda/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/conda/include/python3.7m -I/etc/pyFFTW/include -I/etc/pyFFTW/pyfftw -I/opt/conda/lib/python3.7/site-packages/numpy/core/include -I/opt/conda/include -I/opt/conda/include/python3.7m -c pyfftw/pyfftw.c -o build/temp.linux-aarch64-3.7/pyfftw/pyfftw.o
In file included from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1832:0,
                 from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from pyfftw/pyfftw.c:606:
/opt/conda/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it with " \
  ^~~~~~~
pyfftw/pyfftw.c: In function ‘__pyx_f_6pyfftw_6pyfftw_byte_align.isra.64.constprop’:
pyfftw/pyfftw.c:3969:6: warning: ‘__pyx_v_update_dtype’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   if (__pyx_t_1) {
      ^
creating build/lib.linux-aarch64-3.7
creating build/lib.linux-aarch64-3.7/pyfftw
gcc -pthread -shared -B /opt/conda/compiler_compat -L/opt/conda/lib -Wl,-rpath=/opt/conda/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-aarch64-3.7/pyfftw/pyfftw.o -L/opt/conda/lib -lfftw3f -lm -o build/lib.linux-aarch64-3.7/pyfftw/pyfftw.cpython-37m-aarch64-linux-gnu.so
copying build/lib.linux-aarch64-3.7/pyfftw/pyfftw.cpython-37m-aarch64-linux-gnu.so -> pyfftw
/opt/conda/lib/python3.7/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /etc/pyFFTW/pyfftw/pyfftw.pxd
  tree = Parsing.p_module(s, pxd, full_module_name)
warning: pyfftw/pyfftw.pyx:344:22: Casting a GIL-requiring function into a nogil function circumvents GIL validation
warning: pyfftw/pyfftw.pyx:364:23: Casting a GIL-requiring function into a nogil function circumvents GIL validation
Removing intermediate container 7565f680b6fd
 ---> 409c1e190535
```

Output:
a = pyfftw.empty_aligned(4, dtype='complex128')
fft = pyfftw.builders.fft(a)
a[:] = [1, 2, 3, 4]
fft() #

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
RuntimeError: Undefined plan with nthreads. This is a bug

Exception ignored in: 'pyfftw.pyfftw._fftw_plan_with_nthreads_null'
RuntimeError: Undefined plan with nthreads. This is a bug

array([10.+0.j, -2.+2.j, -2.+0.j, -2.-2.j], dtype=complex64)

**Comment by hgomersall** [comment]
Ach, you need `--enable-threads` or `-enable-openmp` when building FFTW.

**Comment by hgomersall** [comment]
Also, from what I remember, you only build one precision at a time, so you need to run it three times to build all the precisions. It also seems the flag for single precision is [`--enable-float`](http://www.fftw.org/fftw3_doc/Installation-on-Unix.html#Installation-on-Unix).

**Comment by masip85** [comment]
> Ach, you need `--enable-threads` or `--enable-openmp` when building FFTW.

why? I don't want to parallelize,just a single core!

**Comment by hgomersall** [comment]
> why? I don't want to parallelize,just a single core!

So it builds and runs. Patches are welcome that fix the case where missing the missing threads library causes it to not work with a single thread.

**Comment by masip85** [comment]
But one question, if I compile enabling threads,does it force it to use threads when I use it? Or I am just adding a possibility?

**Comment by hgomersall** [comment]
> But one question, if I compile enabling threads,does it force it to use threads when I use it? Or I am just adding a possibility?

It will use as many threads as you request.

**Comment by masip85** [comment]
Ok, so the normal way pyfft needs fftw,is with thread enabled characteristic , even if at the end you use one core. So..I guess you don't know, but, why isn't this a default?
And you compile three times, but for neon support for example there are only two possible configurations:

```
Configure settings for single**: --enable-single** --enable-neon --enable-armv7a-cycle-counter ARM_CPU_TYPE=cortex-a9
Configure settings for double: --enable-armv7a-cycle-counter ARM_CPU_TYPE=cortex-a9
```

I think this is not a problem.

**Comment by hgomersall** [comment]
> So..I guess you don't know, but, why isn't this a default?

Because it's easier to design and maintain to only have one code path. In particular, the library abstraction architecture would get more complicated.

**Comment by masip85** [comment]
What I wanted to say, is that fftw should be always with threads enabled. Less options & less paths.

**Comment by hgomersall** [comment]
> What I wanted to say, is that fftw should be always with threads enabled. Less options & less paths.

I guess it's a low level library that might not always have a threading library available.

**Comment by masip85** [comment]
Well, with threads enables this issue is solved:


I have used single precision in fftw config and float32 numpy:
![image](https://user-images.githubusercontent.com/12863854/87417052-e5a63a80-c5cf-11ea-8081-01e6cd8b5941.png)

With float64 numpy:
![image](https://user-images.githubusercontent.com/12863854/87416864-9fe97200-c5cf-11ea-9312-dfd60bc04901.png)

**Comment by hgomersall** [comment]
>  Normal behaviour in absence of fftw double precision compiled file. Maybe,instead of a warning it should give an error.

It also looks like the wrong answer. Are you able to try calling `fft` with `a` passed in explicitly to check it is actually calculating the result.

I think a warning is the correct behaviour. Presumably you _want_ the answer? The warning is to say the precision might not be what you expect.

**Comment by danschef** [comment]
I am now facing the same issue but only with the latest pyfftw release (0.13.0). I am using the pyfftw version from conda-forge installed into a mambaforge environment. Here are the steps to reproduce:

- install [Mambaforge ](https://github.com/conda-forge/miniforge) (any other conda should also do it)
- create an environment: `mamba create -n debugenv 'python=3.9' numpy 'pyfftw=0.13.0'`
- conda activate debugenv
- in a python console:


```python
(pyfftw) me:~$ python
Python 3.9.9 | packaged by conda-forge | (main, Dec 20 2021, 02:40:17)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> import pyfftw
>>> print(pyfftw.__version__)
0.13.0
>>> arr = np.random.randint(1, 500, (128, 128)).astype(np.complex64)
>>> print("arr:", arr)
arr: [[369.+0.j 221.+0.j 382.+0.j ... 495.+0.j 407.+0.j 254.+0.j]
 [147.+0.j 329.+0.j 367.+0.j ... 337.+0.j  90.+0.j 210.+0.j]
 [207.+0.j 277.+0.j  93.+0.j ... 460.+0.j 162.+0.j  94.+0.j]
 ...
 [377.+0.j 487.+0.j  45.+0.j ...  14.+0.j 425.+0.j 446.+0.j]
 [176.+0.j  87.+0.j  66.+0.j ... 426.+0.j 329.+0.j 263.+0.j]
 [123.+0.j  97.+0.j 309.+0.j ... 328.+0.j 330.+0.j 228.+0.j]]
>>> out = pyfftw.FFTW(arr, np.empty_like(arr), axes=(0, 1))()
RuntimeError: Undefined plan with nthreads. This is a bug
Exception ignored in: 'pyfftw.pyfftw._fftw_plan_with_nthreads_null'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Undefined plan with nthreads. This is a bug
>>>
>>> print(out)
[[0.+0.j 0.+0.j 0.+0.j ... 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j ... 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j ... 0.+0.j 0.+0.j 0.+0.j]
 ...
 [0.+0.j 0.+0.j 0.+0.j ... 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j ... 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j ... 0.+0.j 0.+0.j 0.+0.j]]

```

The issue is that when I call this code within a script the `RuntimeError: Undefined plan with nthreads. This is a bug` immediately crashes the interpreter without any traceback. Also, there does not seem to be a way to catch the RuntimeError. A try/except clause does not work.

**Comment by grlee77** [comment]
Thanks for reporting this @danschef, I opened https://github.com/conda-forge/pyfftw-feedstock/issues/51 to track this since it seems to be a conda packaging issue. In the meantime, here are three potential workarounds:

1.) Force installation of the prior build (build "0") using:
```
mamba install -c conda-forge "pyfftw=0.13.0=py39h51d1ae8_0"
```

This prior build does have potential symbol clashes with `mkl_fft` if that is also present, but I think if you are using NumPy from conda-forge this is probably not an issue. A user reported it for NumPy from defaults so we tried to go back to static linking of FFTW in build 1, but apparently something has gone wrong with that build.

2.) Revert to the older, 0.12 release from conda-forge:  `mamba install pyfftw=0.12`

3.) uninstall the mamba package and install from pip
```
mamba uninstall pyfftw
pip install pyfftw
```

**Comment by danschef** [comment]
Thanks for taking care of this, @grlee77. I temporarily reverted to 0.12 for now and subscribed to the updates of the feedstock issue.

**Comment by lanougue** [comment]
Hello Guys,
Any news on this issue ?
Thanks !

**Comment by paugier** [comment]
The conda-forge issue seems to be fixed by a new build of 0.13.1. I close but do not hesitate to reopen if needed.

**Comment by paugier** [comment]
@ahmad681 can you please tell us exactly how you can obtain this error so that I can try to reproduce?

How did you install pyfftw and fftw?
